cmd/rocsp-tool/main.go

```diff
@@ -118,31 +154,285 @@ func main2() error {
+	}
+	dsn, err := dbConfig.URL()
+	if err != nil {
+		return nil, fmt.Errorf("loading DB URL: %w", err)
+	}
+
+	conf, err := mysql.ParseDSN(dsn)
+	if err != nil {
+		return nil, fmt.Errorf("while parsing DSN from 'DBConnectFile': %s", err)
+	}
+
+	if len(conf.Params) == 0 {
+		conf.Params = make(map[string]string)
+	}
+	conf.Params["tx_isolation"] = "'READ-UNCOMMITTED'"
+	conf.Params["interpolateParams"] = "true"
+	conf.Params["parseTime"] = "true"
+
+	db, err := sql.Open("mysql", conf.FormatDSN())
+	if err != nil {
+		return nil, fmt.Errorf("couldn't setup database client: %s", err)
+	}
+
+	db.SetMaxOpenConns(dbConfig.MaxOpenConns)
+	db.SetMaxIdleConns(dbConfig.MaxIdleConns)
+	db.SetConnMaxLifetime(dbConfig.ConnMaxLifetime.Duration)
+	db.SetConnMaxIdleTime(dbConfig.ConnMaxIdleTime.Duration)
+	return db, nil
+}
+
+// setDefault sets the target to a default value, if it is zero.
+func setDefault(target *int, def int) {
+	if *target == 0 {
+		*target = def
+	}
+}
+
+type client struct {
+	issuers       []ShortIDIssuer
+	redis         *rocsp.WritingClient
+	db            *sql.DB // optional
+	ocspGenerator capb.OCSPGeneratorClient
+	clk           clock.Clock
+}
+
+func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
+	// To scan the DB efficiently, we want to select only currently-valid certificates. There's a
+	// handy expires index, but for selecting a large set of rows, using the primary key will be
+	// more efficient. So first we find a good id to start with, then scan from there. Note: since
+	// AUTO_INCREMENT can skip around a bit, we add padding to ensure we get all currently-valid
+	// certificates.
+	startTime := cl.clk.Now().Add(-24 * time.Hour)
+	var minID int64
+	err := cl.db.QueryRowContext(
+		ctx,
+		"SELECT MIN(id) FROM certificateStatus WHERE notAfter >= ?",
+		startTime,
+	).Scan(&minID)
+	if err != nil {
+		return fmt.Errorf("selecting minID: %w", err)
+	}
+
+	setDefault(&speed.RowsPerSecond, 2000)
+	setDefault(&speed.ParallelSigns, 100)
+
+	query := fmt.Sprintf("SELECT %s FROM certificateStatus WHERE id > ?",
+		strings.Join(sa.CertStatusMetadataFields(), ", "))
+	rows, err := cl.db.QueryContext(ctx, query, minID)
+	if err != nil {
+		return fmt.Errorf("scanning certificateStatus: %w", err)
+	}
+	defer rows.Close()
+
+	// Limit the rate of reading rows.
+	frequency := time.Duration(float64(time.Second) / float64(time.Duration(speed.RowsPerSecond)))
+	rowTicker := time.NewTicker(frequency)
+
+	statusesToSign := make(chan *sa.CertStatusMetadata)
+	successes := make(chan struct{}, speed.ParallelSigns)
+	errChan := make(chan error, speed.ParallelSigns)
+
+	var signersWg sync.WaitGroup
+
+	for i := 0; i < speed.ParallelSigns; i++ {
+		signersWg.Add(1)
+		go cl.signAndStoreResponses(statusesToSign, errChan, successes, &signersWg)
+	}
+	var n, errorCount int
+	// Consume all available successes and errors from the output channels.
+	// If none are immediately available, return.
+	consume := func() {
```

Comment:
Breaking `consume()` out into an anonymous function like this doesn't seem to make the loop below that much more readable to me. Particularly since it is only called in once place, I'm not sure of the value of this layer of indirection/abstraction?

Reply:
Good catch. This was part of an earlier iteration where I had to call it in two places. I'll defactor this.